### PR TITLE
refactor: Simplify / unify material writers

### PR DESCRIPTION
**Pull request by paulgessinger**
This removes the complicated templated `MaterialWriterT` and makes material writers implement `IMaterialWriter` directly.

This is rebased on top of #915, and is WIP until that is merged.

### CODE REVIEW

File: Plugins/Json/src/MaterialMapJsonConverter.cpp [hunk 1]
```diff
@@ -321,9 +318,19 @@ void Acts::MaterialMapJsonConverter::convertToHierarchy(
...
+
+            std::pair p{id, sm};
+            surfaceHierarchy.push_back(p);
+
+            // surfaceHierarchy.push_back(
+            //     std::pair{ssf->geometryId(), defaultSurfaceMaterial(sp)});
+
+            // surfaceHierarchy.push_back(
+            //     {ssf->geometryId(),
+            //      defaultSurfaceMaterial(ssf->getSharedPtr())});
```
Comment (Corentin-Allaire) [inline Plugins/Json/src/MaterialMapJsonConverter.cpp:333]: I guess you wrote those 3 to compare which look the best ? You should remove the one you commented out probably

File: Examples/Run/Vertexing/TrackReaderVertexingPerformanceWriterExample.cpp [hunk 2]
```diff
@@ -61,11 +61,10 @@ int main(int argc, char* argv[]) {
 
   RootParticleReader::Config particleReaderConfig;
   particleReaderConfig.particleCollection = "allTruthParticles";
-  particleReaderConfig.inputFile = "particles_initial.root";
+  particleReaderConfig.filePath = inputDir + "/particles_initial.root";
```
Comment (Corentin-Allaire) [inline Examples/Run/Vertexing/TrackReaderVertexingPerformanceWriterExample.cpp:64]: As I said in #915 I think inputFilePath would look better. But I leave that up to you.

**Review by Corentin-Allaire — COMMENTED** [review]
Look good I just have some tiny comments but I think this can go in after #915

File: Plugins/Json/src/MaterialMapJsonConverter.cpp (hunk 1 above)
Comment (paulgessinger) [reply Plugins/Json/src/MaterialMapJsonConverter.cpp:333]: Oh yeah, my bad.

Comment (paulgessinger) [reply Plugins/Json/src/MaterialMapJsonConverter.cpp]: Should be fixed

**Review by Corentin-Allaire — APPROVED** [review]

**Comment by paulgessinger** [comment]
#915 is merged, so this one is ready to go.

**Pull request merged**